File: app/src/main/java/wangdaye/com/geometricweather/common/basic/models/weather/Current.java
package wangdaye.com.geometricweather.common.basic.models.weather;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.io.Serializable;

import wangdaye.com.geometricweather.common.basic.models.options.unit.DistanceUnit;
import wangdaye.com.geometricweather.common.basic.models.options.unit.RelativeHumidityUnit;
import wangdaye.com.geometricweather.common.basic.models.options.unit.TemperatureUnit;

/**
 * Current.
 *
 * default unit
 * {@link #relativeHumidity} : {@link RelativeHumidityUnit#PERCENT}
 * {@link #dewPoint} : {@link TemperatureUnit#C}
 * {@link #visibility} : {@link DistanceUnit#KM}
 * {@link #ceiling} : {@link DistanceUnit#KM}
 * */
public class Current implements Serializable {

    @Nullable private final String weatherText;
    @Nullable private final WeatherCode weatherCode;

    @Nullable private final Temperature temperature;
    @Nullable private final Precipitation precipitation;
    @Nullable private final Wind wind;
    @Nullable private final UV uv;
    @Nullable private final AirQuality airQuality;

    @Nullable private final Float relativeHumidity;
    @Nullable private final Float pressure;
    @Nullable private final Float visibility;
    @Nullable private final Integer dewPoint;
    @Nullable private final Integer cloudCover;
    @Nullable private final Float ceiling;

    @Nullable private final String dailyForecast;
    @Nullable private final String hourlyForecast;

    public Current(@Nullable String weatherText, @Nullable WeatherCode weatherCode,
                   @Nullable Temperature temperature, @Nullable Precipitation precipitation,
                   @Nullable Wind wind, @Nullable UV uv, @Nullable AirQuality airQuality,
                   @Nullable Float relativeHumidity, @Nullable Float pressure, @Nullable Float visibility,
                   @Nullable Integer dewPoint, @Nullable Integer cloudCover, @Nullable Float ceiling,
                   @Nullable String dailyForecast, @Nullable String hourlyForecast) {
        this.weatherText = weatherText;
        this.weatherCode = weatherCode;
        this.temperature = temperature;
        this.precipitation = precipitation;
        this.wind = wind;
        this.uv = uv;
        this.airQuality = airQuality;
        this.relativeHumidity = relativeHumidity;
        this.pressure = pressure;
        this.visibility = visibility;
        this.dewPoint = dewPoint;
        this.cloudCover = cloudCover;
        this.ceiling = ceiling;
        this.dailyForecast = dailyForecast;
        this.hourlyForecast = hourlyForecast;
    }

    @Nullable
    public String getWeatherText() {
        return weatherText;
    }

    @Nullable
    public WeatherCode getWeatherCode() {
        return weatherCode;
    }

    @Nullable
    public Temperature getTemperature() {
        return temperature;
    }

    @Nullable
    public Precipitation getPrecipitation() {
        return precipitation;
    }

    @Nullable
    public Wind getWind() {
        return wind;
    }

    @Nullable
    public UV getUV() {
        return uv;
    }

    @Nullable
    public AirQuality getAirQuality() {
        return airQuality;
    }

    @Nullable
    public Float getRelativeHumidity() {
        return relativeHumidity;
    }

    @Nullable
    public Float getPressure() {
        return pressure;
    }

    @Nullable
    public Float getVisibility() {
        return visibility;
    }

    @Nullable
    public Integer getDewPoint() {
        return dewPoint;
    }

    @Nullable
    public Integer getCloudCover() {
        return cloudCover;
    }

    @Nullable
    public Float getCeiling() {
        return ceiling;
    }

    @Nullable
    public String getDailyForecast() {
        return dailyForecast;
    }

    @Nullable
    public String getHourlyForecast() {
        return hourlyForecast;
    }
}

File: domain/src/main/java/breezyweather/domain/weather/model/Current.kt
/**
 * This file is part of Breezy Weather.
 *
 * Breezy Weather is free software: you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by the
 * Free Software Foundation, version 3 of the License.
 *
 * Breezy Weather is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
 * License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with Breezy Weather. If not, see <https://www.gnu.org/licenses/>.
 */

package breezyweather.domain.weather.model

import breezyweather.domain.weather.reference.WeatherCode
import breezyweather.domain.weather.wrappers.CurrentWrapper
import org.breezyweather.unit.distance.Distance
import org.breezyweather.unit.pressure.Pressure
import java.io.Serializable

/**
 * Current.
 *
 * default unit
 * [.relativeHumidity] : [RelativeHumidityUnit.PERCENT]
 * [.dewPoint] : [TemperatureUnit.C]
 */
data class Current(
    val weatherText: String? = null,
    val weatherCode: WeatherCode? = null,
    val temperature: Temperature? = null,
    val wind: Wind? = null,
    val uV: UV? = null,
    val airQuality: AirQuality? = null,
    val relativeHumidity: Double? = null,
    val dewPoint: Double? = null,
    /**
     * Pressure at sea level
     * Use Kotlin extensions to initialize this value, like 1013.25.hectopascals
     */
    val pressure: Pressure? = null,
    val cloudCover: Int? = null,
    val visibility: Distance? = null,
    val ceiling: Distance? = null,
    val dailyForecast: String? = null,
    // Is actually a description of the nowcast
    val hourlyForecast: String? = null,
) : Serializable {

    fun toCurrentWrapper() = CurrentWrapper(
        weatherText = this.weatherText,
        weatherCode = this.weatherCode,
        temperature = this.temperature?.toTemperatureWrapper(),
        wind = this.wind,
        uV = uV ?: this.uV,
        relativeHumidity = this.relativeHumidity,
        dewPoint = this.dewPoint,
        pressure = this.pressure,
        cloudCover = this.cloudCover,
        visibility = this.visibility,
        ceiling = this.ceiling,
        dailyForecast = this.dailyForecast,
        hourlyForecast = this.hourlyForecast
    )
}
